### PR TITLE
[Feature] Support DFlash Speculative Decoding Training for Qwen3.5 Models

### PR DESCRIPTION
**Pull request by EanWang211123**
## PR Summary

This PR adds support for **DFlash draft model speculative decoding training** for the **Qwen/Qwen3.5 series**. It includes new features and VLM support for these models. This update enables users to train draft models with **Qwen3.5-27B** and **Qwen3.5-35B-A3B** using **HF backend** with full compatibility for VLM-based configurations. Some parts remain limited to **TP=1** and **HF backend** for now.

---

## Motivation

This PR enables **DFlash speculative decoding training** for the following models:

* **Qwen/Qwen3.5-27B**
  [https://huggingface.co/Qwen/Qwen3.5-27B](https://huggingface.co/Qwen/Qwen3.5-27B)

* **Qwen/Qwen3.5-35B-A3B**
  [https://huggingface.co/Qwen/Qwen3.5-35B-A3B](https://huggingface.co/Qwen/Qwen3.5-35B-A3B)

### Current Limitations:

* **Only supports HF backend** for online training.
* **Does not support TP or SGLang backend**.
* SGLang support depends on a future **SGLang release** for Qwen3.5 compatibility.

The **DFlash hidden-layer extraction logic** remains unchanged:

* **Only full-attention layers** are used.
* **Five layers are evenly selected** as draft layers.

---

### `scripts/train_dflash.py`

* Added VLM-related options: `--is-vlm`, `--text-only`, `--min-pixels`, `--max-pixels`
* Added new helper functions: `_resolve_target_num_hidden_layers`, `_resolve_draft_config`, `_resolve_target_weight_keys`, `_build_target_layer_ids` to handle VLM `text_config`, number of layers, weight paths, and `target_layer_ids`
* `build_models`: Uses `Qwen3_5ForConditionalGeneration` / `Qwen3_5MoeForConditionalGeneration` to load Qwen3.5 models
* `build_dataloader`: Added `is_vlm` and `processor` to support VLM data
* `main`: VLM detection, processor loading, and passing `pixel_values` and other VLM-related inputs to the target model

### `specforge/modeling/target/dflash_target_model.py`

* Defined `QWEN3_5_MODEL_TYPES` and `VLM_MODEL_TYPES`
* `DFlashTargetOutput`: Added `position_ids`; `generate_dflash_data`: Added `pixel_values`, `image_grid_thw`, and `video_grid_thw`
* `HFDFlashTargetModel`: Branches based on `model_type` to load Qwen3.5 models; when VLM is enabled, passes multimodal parameters during `forward`

### `specforge/modeling/target/target_utils.py`

* Added `_get_text_config_for_embeddings`: Handles `vocab_size` and `hidden_size` for VLM from `text_config`

### `specforge/core/dflash.py`

* `forward` added `attention_mask` and `position_ids`; introduced `_create_position_ids_from_context` to support multimodal `[3, B, S]` position_ids

### `specforge/data/template.py`

* Added `qwen3.5` and `qwen3.5-thinking` chat templates

### `pyproject.toml`

* Upgraded `transformers` to `>=5.2.0`

### `custom_backend/*.py` (gpt_oss, llama, llama4, phi3)

* Removed `check_model_inputs` decorator to support compatibility with transformers 5.x

### New Files

* `configs/qwen3.5-27b-dflash.json`, `configs/qwen3.5-35b-a3b-dflash.json`
* `examples/run_qwen3.5_27b_dflash_online.sh`, `examples/run_qwen3.5_35b_a3b_dflash_online.sh`

---

## Other Functionality Impact Analysis

* **Other DFlash models**: No impact, still follows the general `AutoModelForCausalLM` path.
* **Eagle3**: No impact, continues to use an independent target module.
* **Non-VLM models**: When `position_ids=None` in `forward`, the original logic is used, and behavior remains unchanged.

---

## Related Issues

[https://github.com/sgl-project/SpecForge/issues/486](https://github.com/sgl-project/SpecForge/issues/486)

---

### Environment

```bash
H100 × 8
```

### Dataset

[https://huggingface.co/datasets/zhaode/EagleChat](https://huggingface.co/datasets/zhaode/EagleChat)

Randomly sampled:

```bash
8k samples
```

---

## Training Command

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
torchrun \
--standalone \
--nproc_per_node 8 \
scripts/train_dflash.py \
--target-model-path /models/Qwen/Qwen3___5-27B/ \
--draft-config-path configs/qwen3.5-27b-dflash.json \
--tp-size 1 \
--is-vlm \
--text-only \
--trust-remote-code \
--train-data-path datasets/qwen3_5_eaglechat_train_mixed_10k_regen.jsonl \
--chat-template qwen3.5 \
--output-dir outputs/qwen3.5-27b-dflash-10k-eagle-test \
--cache-dir cache \
--num-epochs 6 \
--batch-size 1 \
--learning-rate 6e-4 \
--max-length 4096 \
--num-draft-layers 5 \
--block-size 16 \
--num-anchors 512 \
--loss-decay-gamma 7.0 \
--max-length 4096
```

---

## Benchmark Test

Use the following SGLang PR for testing:

[https://github.com/sgl-project/sglang/pull/19952](https://github.com/sgl-project/sglang/pull/19952)

Benchmark script:

[https://github.com/z-lab/dflash](https://github.com/z-lab/dflash)

```bash
SGLANG_DISABLE_CUDNN_CHECK=1 \
CUDA_VISIBLE_DEVICES=4,5,6,7 \
python benchmark_sglang.py \
--tp-size 4 \
--target-model /models/Qwen/Qwen3___5-27B/ \
--draft-model /draft-model-path/ \
--concurrencies 1,8,32 \
--dataset-name humaneval \
--attention-backends fa3 \
--max-running-requests 32
```

---

# Results

### Baseline Output

| concurrency | tok/s   |
| ----------- | ------- |
| 1           | 103.61  |
| 4           | 361.46  |
| 32          | 1860.84 |

### DFlash Output

| concurrency | tok/s   |
| ----------- | ------- |
| 1           | 207.19  |
| 4           | 684.67  |
| 32          | 2115.39 |

### Speedup (DFLASH / Baseline)

| concurrency | speedup |
| ----------- | ------- |
| 1           | 2.000x  |
| 4           | 1.894x  |
| 32          | 1.137x  |

### DFlash Acceptance Length

| concurrency | acceptance length |
| ----------- | ----------------- |
| 1           | 3.233             |
| 4           | 3.216             |
| 32          | 3.219             |

---

# Notes

Current limitations:

* **HF backend**
* **TP = 1**

### Possible OOM Issues

If users encounter **GPU OOM**, they can:

* Enable **FSDP CPUOffload** for draft model training
* Use `device_map="auto"` to reduce memory usage

---

## Future SGLang / TP Backend Support

I’ve implemented **SGLang backend target model support** in my personal validation branch. However, merging into the main repository requires:

* SGLang **official support for Qwen3.5 (≥ 0.5.9)**
* **Stable DFlash interface** support in SGLang

Training can still be performed using:

* **SGLang PR**: [https://github.com/sgl-project/sglang/pull/19952](https://github.com/sgl-project/sglang/pull/19952)
* **My personal validation branch**: [https://github.com/EanWang211123/SpecForge/tree/qwen3_5_online_sglang](https://github.com/EanWang211123/SpecForge/tree/qwen3_5_online_sglang)

This approach:

* **Reduces memory usage significantly**
* **Improves training speed**

---

### Key Files to Review:

1. **`scripts/train_dflash.py`** — contains most of the changes related to the new VLM support and command-line arguments.
2. **`specforge/modeling/target/dflash_target_model.py`** — key changes in how Qwen3.5 models are loaded, and the new multi-modal parameters passed during `forward`.
3. **`specforge/core/dflash.py`** — changes to how attention masks and position IDs are handled, especially for multi-modal inputs.

### Areas of Focus:

* **New VLM functionalities** and how these interact with the draft model.
* **Training and inference commands**: Check if the new configuration works as expected in different environments.
* **Backwards compatibility**: Ensure no regressions for non-VLM models.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by jiapingW** [comment]
Great work. We recently upgraded our sglang version and updated related dependencies. Could you help fix the conflicts?

**Comment by yangshen88-rgb** [comment]
Hello, your work is excellent. I would like to ask if there is training support for EAGLE-3 on the multimodal Qwen-3.5-27B?
